medium-tests/src/test/java/mediumtest/synchronization/PluginSynchronizationMediumTests.java

```diff
@@ -0,0 +1,330 @@
+import org.sonarsource.sonarlint.core.serverconnection.proto.Sonarlint.PluginReferences.PluginReference;
+import org.sonarsource.sonarlint.core.serverconnection.storage.ProtobufFileUtil;
+import testutils.PluginLocator;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static mediumtest.fixtures.ServerFixture.newSonarQubeServer;
+import static mediumtest.fixtures.ServerFixture.ServerStatus.DOWN;
+import static mediumtest.fixtures.ServerFixture.ServerStatus.UP;
+import static mediumtest.fixtures.SonarLintBackendFixture.newBackend;
+import static mediumtest.fixtures.SonarLintBackendFixture.newFakeClient;
+import static org.assertj.core.api.Assertions.as;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.entry;
+import static org.assertj.core.api.InstanceOfAssertFactories.MAP;
+import static org.awaitility.Awaitility.waitAtMost;
+import static org.sonarsource.sonarlint.core.serverconnection.storage.ProjectStoragePaths.encodeForFs;
+
+class PluginSynchronizationMediumTests {
+
+  @Test
+  void it_should_pull_plugins_at_startup_from_the_server() {
+    var server = newSonarQubeServer("10.3")
+      .withStatus(UP)
+      .withPlugin(TestPlugin.JAVA)
+      .withProject("projectKey", project -> project.withBranch("main"))
+      .start();
+    backend = newBackend()
+      .withEnabledLanguageInStandaloneMode(Language.JAVA)
+      .withSonarQubeConnection("connectionId", server)
+      .withBoundConfigScope("configScopeId", "connectionId", "projectKey", "main")
+      .withFullSynchronization()
+      .build();
+
+    waitAtMost(3, SECONDS).untilAsserted(() -> {
+      assertThat(getPluginsStorageFolder()).isDirectoryContaining(path -> path.getFileName().toString().equals("sonar-java-plugin-7.15.0.30507.jar"));
+      assertThat(getPluginReferencesFilePath())
+        .exists()
+        .extracting(this::readPluginReferences, as(MAP))
+        .containsOnly(
+          entry("java", PluginReference.newBuilder().setFilename("sonar-java-plugin-7.15.0.30507.jar").setKey("java").setHash(PluginLocator.SONAR_JAVA_PLUGIN_JAR_HASH).build()));
+    });
+  }
+
+  @Test
+  void it_should_notify_clients_when_plugins_have_been_pulled() {
+    var server = newSonarQubeServer("10.3")
+      .withPlugin(TestPlugin.JAVA)
+      .withProject("projectKey", project -> project.withBranch("main"))
+      .start();
+    var client = newFakeClient().build();
+    backend = newBackend()
+      .withExtraEnabledLanguagesInConnectedMode(Language.JAVA)
+      .withSonarQubeConnection("connectionId", server)
+      .withBoundConfigScope("configScopeId", "connectionId", "projectKey", "main")
+      .withFullSynchronization()
+      .build(client);
+
+    waitAtMost(3, SECONDS).untilAsserted(() -> {
+      assertThat(getPluginsStorageFolder()).isNotEmptyDirectory();
+      assertThat(client.getConnectionsWhichUpdatedPlugins()).containsExactly("connectionId");
```

Comment:
And same for assertions, we can use Mockito (assuming the fakeClient has been spied):
```java
verify(client).didUpdatePlugins(argCaptor.capture());
```

Reply:
Same answer

Reply:
Another thing to consider is that verifying this kind of calls is a bit more verbose as we often have to use a captor